src/lexer.coffee

```diff
@@ -1212,11 +1212,14 @@ CSX_ATTRIBUTE = /// ^
 ///
 
 NUMBER     = ///
-  ^ 0b[01]+n?    |            # binary
-  ^ 0o[0-7]+n?   |            # octal
-  ^ 0x[\da-f]+n? |            # hex
-  ^ \d+n         |            # decimal bigint
-  ^ \d*\.?\d+ (?:e[+-]?\d+)?  # decimal
+  ^ 0b[01](?:_?[01])*                           | # binary
+  ^ 0o[0-7](?:_?[0-7])*                         | # octal
+  ^ 0x[\da-f](?:_?[\da-f])*                     | # hex
+  ^ \d+n                                        | # decimal bigint
+  ^ (?:\d(?:_?\d)*)?    \.?   (?:\d(?:_?\d)*)+    # decimal
+                    (?:e[+-]? (?:\d(?:_?\d)*)+ )?
```

Comment:
Now that I merged in #5104, this regex needs to be updated to handle `BigInt`s with numeric separators.

Also I see two failing tests on this branch, whereas all the tests pass on `master`. This first one passes on `master` but fails here, presumably because of changes to this regex:

```
test/numbers_bigint.coffee:11:19: error: unexpected identifier
eq 42n, 0b101010n
```

This test failed both before and after I merged `master` into this branch:

```
invalid decimal literal separators do not compile
AssertionError [ERR_ASSERTION] [ERR_ASSERTION]: Missing expected exception.
  at doesNotCompile (test/numeric_literal_separators.coffee:1:3)
doesNotCompile '1._23'
```

For some reason `1._23` is allowed by the grammar. [Even in current CoffeeScript this is permitted, and compiles into `1.._23`](https://coffeescript.org/#try:1._23), which seems like a bug. Node seems to allow `1.._23` as valid JavaScript, though if you set `x` to it then `x` remains undefined, so I’m not sure what this code actually represents.